### PR TITLE
Feature/hypothetical protein replacement

### PR DESCRIPTION
**Pull request by tgurbich**
Added replacement of protein function from "hypothetical protein" to something meaningful from UniFIRE, InterPro or eggNOG (if available).

### CODE REVIEW

File: bin/add_hypothetical_protein_descriptions.py [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+def keep_or_move_to_note(found_function, function_source, col9_dict):
+    """
+    Function aims to identify if a description is likely to be a sentence/paragraph rather than
+    a succinct function description. If it's the former, move it to note and revert function to
+    "hypothetical protein".
+    """
+    move_to_note = False
+    description_length = len(found_function.split())
+    if description_length < 12:  # likely to be a proper description
+        text_to_avoid = [
```
Comment (mberacochea) [inline bin/add_hypothetical_protein_descriptions.py:108]: I would push there 2 constants to the top of the file and use upper case  (it's a convention for constants)

File: bin/add_hypothetical_protein_descriptions.py [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+            move_to_note = True
+        if any(
+            found_function.lower().startswith(phrase.lower())
+            for phrase in starts_to_avoid
+        ):
+            move_to_note = True
+        if move_to_note:
+            col9_dict = move_function_to_note(found_function, col9_dict)
+            found_function = "hypothetical protein"
+            function_source = "Prokka"
```
Comment (mberacochea) [inline bin/add_hypothetical_protein_descriptions.py:169]: Do we need to specify the version?

File: bin/add_hypothetical_protein_descriptions.py [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+            "related to",
+        ]
+        if any(phrase.lower() in found_function.lower() for phrase in text_to_avoid):
+            move_to_note = True
+        if any(
+            found_function.lower().startswith(phrase.lower())
+            for phrase in starts_to_avoid
+        ):
+            move_to_note = True
+        if move_to_note:
```
Comment (mberacochea) [inline bin/add_hypothetical_protein_descriptions.py:166]: I slighter different way of doing this:
```python
valid_function = any(phrase.lower() in found_function.lower() for phrase in text_to_avoid)
valid_phrase = any(found_function.lower().startswith(phrase.lower()) for phrase in starts_to_avoid)
move_to_note = valid_function or valid_phrase 
```

File: bin/add_hypothetical_protein_descriptions.py [hunk 4]
```diff
@@ -0,0 +1,714 @@
...
+        and "protein" not in found_function.lower()
+    ):
+        found_function += " protein"
+    if (
+        found_function.lower().endswith("binding")
+        and "protein" not in found_function.lower()
+        and "domain" not in found_function.lower()
+    ):
+        found_function += " protein"
+    return found_function
```
Comment (mberacochea) [inline bin/add_hypothetical_protein_descriptions.py:202]:
```suggestion
def clean_up_function(found_function):
    found_function_lower = found_function.lower()
    if "domain" in found_function_lower:
        found_function = reformat_domain(found_function)
    if (
        found_function_lower.endswith("family")
        and "protein" not in found_function.lower()
    ):
        found_function += " protein"
    if (
        found_function_lower.endswith("binding")
        and "protein" not in found_function_lower
        and "domain" not in found_function_lower
    ):
        found_function += " protein"
    return found_function
```

File: bin/add_hypothetical_protein_descriptions.py [hunk 5]
```diff
@@ -0,0 +1,714 @@
...
+        ):
+            if ipr_dict[db]["level"] > best_level or (
+                ipr_dict[db]["level"] == best_level
+                and ipr_dict[db]["match"] > best_fraction
+            ):
+                best_fraction = ipr_dict[db]["match"]
+                best_level = ipr_dict[db]["level"]
+                highest_match = dict()
+                highest_match[db] = ipr_dict[db]
+    if "Pfam" not in highest_match and best_fraction > 0.30:
```
Comment (mberacochea) [inline bin/add_hypothetical_protein_descriptions.py:356]: I would put the thresholds as constants at the top of the file, making it easier to stop and modify if needed.

File: bin/add_hypothetical_protein_descriptions.py [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+    eggnog_info = dict()
+    with open(file, "r") as file_in:
+        for line in file_in:
+            if not line.startswith("#"):
+                cols = line.strip().split("\t")
+                try:
+                    evalue = float(cols[2])
+                except:
+                    continue
+                if evalue > 1e-10:
```
Comment (mberacochea) [inline bin/add_hypothetical_protein_descriptions.py:400]: I would also push threshold as a constant to the top of the file (or the method)

File: bin/annotate_gff.py [hunk 7]
```diff
@@ -45,6 +51,12 @@ def get_eggnog(eggnog_annot):
             if line.startswith("#"):
                 eggnog_fields = get_eggnog_fields(line)
             else:
+                try:
+                    evalue = float(cols[2])
+                except ValueError:
+                    continue
+                if evalue > 1e-10:
```
Comment (mberacochea) [inline bin/annotate_gff.py:58]: Should we print that this entry was ignored because of this in the error log?

**Review by mberacochea — APPROVED** [review]
I did a quick review, I've added some minor comments around the add_hypothetical_protein_descriptions.py script. That script does a lot, so we need to add a few unit tests for it.

File: bin/add_hypothetical_protein_descriptions.py (hunk 1 above)
Comment (tgurbich) [reply bin/add_hypothetical_protein_descriptions.py:113]: thanks, changed that

**Pull request merged**